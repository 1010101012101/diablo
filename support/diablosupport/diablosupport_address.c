/* This research is supported by the European Union Seventh Framework Programme (FP7/2007-2013), project ASPIRE (Advanced  Software Protection: Integration, Research, and Exploitation), under grant agreement no. 609734; on-line at https://aspire-fp7.eu/. */

/* The development of portions of the code contained in this file was sponsored by Samsung Electronics UK. */

#include <diablosupport.h>

t_string
IoModifierAddressGeneric (t_const_string modifiers, va_list * ap)
{
  char buffer[20];
  t_string ret;
  t_string_array *array;
  t_address_generic addr = va_arg (*ap, t_address_generic);

  array = StringArrayNew ();
  if (addr.type == ADDRSIZE32)
    sprintf (buffer, "0x%x", addr.types.addr32);
  else if (addr.type == ADDRSIZE64)
    sprintf (buffer, "0x%"PRIx64"", addr.types.addr64);
  else
    FATAL(("Address not correctly initialized!"));

  StringArrayAppendString (array, StringDup (buffer));

  ret = StringArrayJoin (array, " ");
  StringArrayFree (array);
  return ret;
}

t_string
IoModifierAddress32 (t_const_string modifiers, va_list * ap)
{
  char buffer[20];
  t_string ret;
  t_string_array *array;
  t_uint32 addr = va_arg (*ap, t_uint32);

  array = StringArrayNew ();
  sprintf (buffer, "0x%x", addr);

  StringArrayAppendString (array, StringDup (buffer));

  ret = StringArrayJoin (array, " ");
  StringArrayFree (array);
  return ret;
}

t_string
IoModifierAddress64 (t_const_string modifiers, va_list * ap)
{
  char buffer[20];
  t_string ret;
  t_string_array *array;
  t_uint64 addr = va_arg (*ap, t_uint64);

  array = StringArrayNew ();
  sprintf (buffer, "0x%"PRIx64"", addr);

  StringArrayAppendString (array, StringDup (buffer));

  ret = StringArrayJoin (array, " ");
  StringArrayFree (array);
  return ret;
}

t_address_generic
RealAddressNew32 (t_const_string file, t_uint32 line, t_uint32 a)
{
  t_address_generic ret;

  ret.type = ADDRSIZE32;
  ret.types.addr32 = a;
  return ret;
}

t_address_generic
RealAddressNew64 (t_const_string file, t_uint32 line, t_uint64 a)
{
  t_address_generic ret;

  ret.type = ADDRSIZE64;
  ret.types.addr64 = a;
  return ret;
}

t_address_generic
RealAddressAdd (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s %d %d", line,
          file, a.type, b.type));

  ret.type = b.type;

  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 + b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 + b.types.addr64;
  return ret;
}

t_address_generic
RealAddressAnd (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = b.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 & b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 & b.types.addr64;
  return ret;
}

t_address_generic
RealAddressOr (t_const_string file, t_uint32 line, t_address_generic a,
               t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = a.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 | b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 | b.types.addr64;
  return ret;
}

t_address_generic
RealAddressXor (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = a.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 ^ b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 ^ b.types.addr64;
  return ret;
}

t_address_generic
RealAddressNot (t_const_string file, t_uint32 line, t_address_generic a)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = ~a.types.addr32;
  else
    ret.types.addr64 = ~a.types.addr64;
  return ret;
}

t_address_generic
RealAddressShl (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = b.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 << b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 << b.types.addr64;
  return ret;
}

t_address_generic
RealAddressShr (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));

  ret.type = b.type;
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
    ret.types.addr32 = a.types.addr32 >> b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 >> b.types.addr64;
  return ret;
}

t_address_generic
RealAddressAddInt32 (t_const_string file, t_uint32 line, t_address_generic a,
                     t_int32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 + b;
  else
    ret.types.addr64 = a.types.addr64 + b;
  return ret;
}

t_address_generic
RealAddressAddUint64 (t_const_string file, t_uint32 line, t_address_generic a,
                      t_uint64 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 + (t_uint32)b;
  else
    ret.types.addr64 = a.types.addr64 + b;
  return ret;
}

t_address_generic
RealAddressAddUint32 (t_const_string file, t_uint32 line, t_address_generic a,
                      t_uint32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 + b;
  else
    ret.types.addr64 = a.types.addr64 + b;
  return ret;
}

t_address_generic
RealAddressInverseMaskUint32 (t_const_string file, t_uint32 line, t_address_generic a,
                              t_uint32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 & (~((t_uint32) (b)));
  else
    ret.types.addr64 = a.types.addr64 & (~((t_uint64) (b)));
  return ret;
}

t_address_generic
RealAddressAddInt64 (t_const_string file, t_uint32 line, t_address_generic a,
                     t_int64 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 + (t_int32)b;
  else
    ret.types.addr64 = a.types.addr64 + b;
  return ret;
}

t_address_generic
RealAddressSubUint32 (t_const_string file, t_uint32 line, t_address_generic a,
                      t_uint32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 - b;
  else
    ret.types.addr64 = a.types.addr64 - b;
  return ret;
}

t_address_generic
RealAddressSubInt32 (t_const_string file, t_uint32 line, t_address_generic a,
                     t_int32 b)
{
  t_address_generic ret;

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 - b;
  else
    ret.types.addr64 = a.types.addr64 - b;
  return ret;
}

t_address_generic
RealAddressSub (t_const_string file, t_uint32 line, t_address_generic a,
                t_address_generic b)
{
  t_address_generic ret;

  ASSERT((a.type == b.type),
         ("Sub'ed different size addresses at line %d in %s", line, file));

  ret.type = a.type;
  if (a.type == ADDRSIZE32)
    ret.types.addr32 = a.types.addr32 - b.types.addr32;
  else
    ret.types.addr64 = a.types.addr64 - b.types.addr64;
  return ret;
}

t_bool
RealAddressIsNull (t_const_string file, t_uint32 line, t_address_generic a)
{
  if (a.type == ADDRSIZE32)
    return a.types.addr32 == 0;
  else
    return a.types.addr64 == ULL(0);
}

t_bool
RealAddressIsMax (t_const_string file, t_uint32 line, t_address_generic a)
{
  if (a.type == ADDRSIZE32)
    return a.types.addr32 == 0xFFFFFFFF;
  else
    return a.types.addr64 == ULL(0xFFFFFFFFFFFFFFFF);
}

t_bool
RealAddressIsGt (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));
  if (a.type == ADDRSIZE32)
    return a.types.addr32 > b.types.addr32;
  else
    return a.types.addr64 > b.types.addr64;
}

t_bool
RealAddressIsGe (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type), ("Added different size addresses"));
  if (a.type == ADDRSIZE32)
    return a.types.addr32 >= b.types.addr32;
  else
    return a.types.addr64 >= b.types.addr64;
}

t_bool
RealAddressIsLt (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type),
         ("Added different size addresses at line %d in %s", line, file));
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
  {
    return a.types.addr32 < b.types.addr32;
  }
  else
  {
    return a.types.addr64 < b.types.addr64;
  }
}

t_bool
RealAddressIsLe (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type), ("Added different size addresses"));
  if ((a.type == ADDRSIZE32) || (b.type == ADDRSIZE32))
  {
    return a.types.addr32 <= b.types.addr32;
  }
  else
  {
    return a.types.addr64 <= b.types.addr64;
  }
}

t_bool
RealAddressIsEq (t_const_string file, t_uint32 line, t_address_generic a,
                 t_address_generic b)
{
  ASSERT((a.type == b.type),
         ("AddressEQ at line %d in %s has different type operands", line,
          file));
  if (a.type == ADDRSIZE32)
    return a.types.addr32 == b.types.addr32;
  else
    return a.types.addr64 == b.types.addr64;
}

void *
RealAddressAddDispl (t_const_string file, t_uint32 line, void *in,
                     t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return ((void *) (((char *) in) + addr.types.addr32));
  }
  else
  {
    return ((void *) (((char *) in) + addr.types.addr64));
  }
}

t_address_generic
RealAddressMulUint32 (t_const_string file, t_uint32 line, t_address_generic address,
                      t_uint32 div)
{
  t_address_generic ret;

  ret.type = address.type;
  if (address.type == ADDRSIZE32)
  {
    ret.types.addr32 = address.types.addr32 * div;
  }
  else if (address.type == ADDRSIZE64)
  {
    ret.types.addr64 = address.types.addr64 * ((t_uint64) div);
  }
  else
  {
    FATAL(("Address is corrupt (called at file %s, line %d) ! %d", file,
           line, address.type));
  }
  return ret;
}

t_address_generic
RealAddressDivUint32 (t_const_string file, t_uint32 line, t_address_generic address,
                      t_uint32 div)
{
  t_address_generic ret;

  ret.type = address.type;
  if (address.type == ADDRSIZE32)
  {
    ret.types.addr32 = address.types.addr32 / div;
  }
  else if (address.type == ADDRSIZE64)
  {
    ret.types.addr64 = address.types.addr64 / ((t_uint64) div);
  }
  else
  {
    FATAL(("Address is corrupt (called at file %s, line %d) ! %d", file,
           line, address.type));
  }
  return ret;
}

t_uint32
RealAddressExtractUint32 (t_const_string file, int line, t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return addr.types.addr32;
  }
  if (addr.types.addr64 > 0xffffffffull)
  {
    /* check for overflow: assume there is no overflow if it is a negative
     * number that also fits in 32 bits */
    if ((addr.types.addr64 & 0xffffffff80000000ull) != 0xffffffff80000000ull)
      FATAL(("Overflow at line %d in file %s val = %"PRIx64"\n", line, file, addr.types.addr64));
  }
  return (t_uint32) addr.types.addr64;
}

t_uint64
RealAddressExtractUint64 (t_const_string file, int line, t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return (t_uint64) addr.types.addr32;
  }
  return addr.types.addr64;
}

t_int32
RealAddressExtractInt32 (t_const_string file, int line, t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return (t_int32) addr.types.addr32;
  }
  if (addr.types.addr64 > 0xffffffffull)
  {
    if ((addr.types.addr64 & 0xffffffff80000000ull) !=
        0xffffffff80000000ull)
      FATAL(("Overflow at line %d in file %s\n", line, file));
    else
      /* negative number */
      return (t_int32) ((t_int64) addr.types.addr64);
  }
  return (t_int32) ((t_uint32) addr.types.addr64);
}

t_int64
RealAddressExtractInt64SignExtend (t_const_string file, int line, t_address_generic addr)
{
  if (addr.type == ADDRSIZE32)
  {
    return (t_int32) addr.types.addr32;
  }
  else if (addr.type == ADDRSIZE64)
  {
    return (t_int64) (addr.types.addr64);
  }
  else
  {
    FATAL(("address is corrupt"));
  }
}

t_address_generic
RealAddressSignExtend (t_const_string file, t_uint32 line, t_address_generic x,
                       t_uint32 signbit)
{
  if (x.type == ADDRSIZE32)
  {
    x.types.addr32 = (x.types.addr32 ^ (1 << signbit)) - (1 << signbit);
  }
  else if (x.type == ADDRSIZE64)
  {
    x.types.addr64 = (x.types.addr64 ^ (1 << signbit)) - (1 << signbit);
  }
  else
  {
    FATAL(("address is corrupt"));
  }

  return x;
}

#ifdef _MSC_VER
inline t_address
AddressNewTyped(t_address d, t_uint32 type)
{ 
  t_address res;
  if (type == ADDRSIZE32)
    res = AddressNew32(d);
  else if (type == ADDRSIZE64)
    res = AddressNew64(d);
  else
    FATAL(("Unknown address type (%d)", type));
  return res;   
}

inline t_uint32
WordAlignedSize(t_uint32 s)
{
  t_uint32 res;
  if ((s) <= 8)
    res = 8;
  else if ((s) <= 16)
    res = 16;
  else if ((s) <= 32)
    res = 32;
  else if ((s) <= 64)
    res = 64;
  else
    FATAL(("Unsupported size: %d", s));
  return res;
}
#endif

t_uint32 AddressSizeInBytes(t_address_type address_type)
{
  return (address_type == ADDRSIZE32) ? 4 : 8;
}

/* vim: set shiftwidth=2 expandtab cinoptions=p5,t0,(0, foldmethod=marker tw=80 cindent: */
