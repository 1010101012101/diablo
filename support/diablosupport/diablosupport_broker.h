#include <diablosupport.h>
#ifdef DIABLOSUPPORT_FUNCTIONS
#ifndef DIABLOSUPPORT_BROKER_FUNCTIONS
#define DIABLOSUPPORT_BROKER_FUNCTIONS
t_bool DiabloBrokerCallExists (const char *name);
void DiabloBrokerCallInstall (const char *name, const char *prototype, void *implementation, t_bool final, ...);
t_bool DiabloBrokerCall (const char *name, ...);
void DiabloBrokerFini ();
#endif
#endif
/* vim: set shiftwidth=2 expandtab cinoptions=p5,t0,(0, foldmethod=marker tw=80 cindent: */
