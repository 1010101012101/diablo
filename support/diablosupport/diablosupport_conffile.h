/* This research is supported by the European Union Seventh Framework Programme (FP7/2007-2013), project ASPIRE (Advanced  Software Protection: Integration, Research, and Exploitation), under grant agreement no. 609734; on-line at https://aspire-fp7.eu/. */

#include <diablosupport.h>

#ifdef DIABLOSUPPORT_FUNCTIONS
#ifndef DIABLOSUPPORT_CONFFILE_FUNCTIONS
#define DIABLOSUPPORT_CONFFILE_FUNCTIONS
void ConfValueFree(const void * tf, void *unused_data);
t_string ConfValueGet(t_hash_table * in,t_string x);
t_hash_table * ConfFileRead(t_const_string conffile);
void ConfValuePrint(const void * string1, void * string2);
void ConfValueSet(t_string x, t_string y);
#endif
#endif
