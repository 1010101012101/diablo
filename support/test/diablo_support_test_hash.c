#include <diablosupport.h>
#include <string.h>

void StatusIo(t_uint32 id, t_string out)
{
	printf(out);
}

char *keys[] = {
	NULL,
	"Search",
	"seek",
	"and",
	"know",
	"how",
	"this",
	"foul",
	"murder",
	"comes",
	"-----",
	"Who",
	"seeks",
	"and",
	"will",
	"not",
	"take",
	"when",
	"once",
	"'tis",
	"offer'd",
	"Shall",
	"never",
	"find",
	"it",
	"more",
	"-----",
	"Search",
	"seek",
	"and",
	"know",
	"how",
	"this",
	"foul",
	"murder",
	"comes",
	NULL
};

typedef struct _test1
{
	t_hash_table_node node;
	char *string2;
} hntest1;

typedef struct _test2
{
	char *string2;
	t_hash_table_node node;
} hntest2;

	void
SP (void *in, void *ea)
{
	hntest1 *p = in;
	printf ("p->string=%s\n", p->string2);
}

	void
SP2 (void *in, void *ea)
{
	hntest2 *p = in;
	printf ("p->string=%s\n", p->string2);
}

#ifndef offsetof
#define offsetof(TYPE, MEMBER) ((size_t) &((TYPE *)0)->MEMBER)
#endif

int main()
{
	char **it = keys;
	hntest1 *insert;
	hntest2 *insert2;
	t_hash_table *test1;


	IoHandlerAdd(E_STATUS,StatusIo);
	IoHandlerAdd(E_FATAL,StatusIo);
	IoModifierAdd('%',0,NULL,IoModifierPrintf);



	printf ("Creating simple hash table\n");
	/* We use small hash table size to allow thourough testing */
	test1 = HashTableNew (3, 0, (t_hash_func) StringHash, (t_int32 (*) (void *, void *)) StringCmp, NULL);
	printf ("Done\n");

	it++;
	printf ("Inserting elements\n");
	do
	{
		insert = Malloc (sizeof (hntest1));
		printf ("   Insert: %s\n", *it);
		HASH_TABLE_NODE_SET_KEY(&insert->node,  *it);
		insert->string2 = StringDup (*it);
		HashTableInsert (test1, insert);
	}
	while (*(++it));

	it--;
	do
	{
		printf ("   Lookup: %s\n", *it);
		insert = HashTableLookup (test1, *it);
		printf ("   Verify: %s\n", insert->string2);
		if (strcmp (insert->string2, *it))
		{
			printf ("Failed!\n");
			exit (-1);
		}
	}
	while (*(--it));

	HashTableWalk (test1, SP, NULL);


	printf ("Prepending elements\n");
	printf ("Creating simple hash table\n");
	/* We use small hash table size to allow thourough testing */
	test1 = HashTableNew (3, 0, (t_hash_func) StringHash, (t_int32 (*) (void *, void *)) StringCmp, NULL);
	printf ("Done\n");

	it++;
	printf ("Inserting elements\n");
	do
	{
		insert = Malloc (sizeof (hntest1));
		printf ("   Prepend: %s\n", *it);
		HASH_TABLE_NODE_SET_KEY(&insert->node,  *it);
		insert->string2 = StringDup (*it);
		HashTablePrepend (test1, insert);
	}
	while (*(++it));

	it--;
	do
	{
		printf ("   Lookup: %s\n", *it);
		insert = HashTableLookup (test1, *it);
		printf ("   Verify: %s\n", insert->string2);
		if (strcmp (insert->string2, *it))
		{
			printf ("Failed!\n");
			exit (-1);
		}
	}
	while (*(--it));

	HashTableWalk (test1, SP, NULL);




	printf ("Creating displaced hash table\n");
	/* We use small hash table size to allow thourough testing */
	test1 =
		HashTableNew (3, offsetof (hntest2, node), (t_hash_func) StringHash, (t_int32 (*) (void *, void *)) StringCmp, NULL);
	printf ("Done\n");

	it++;
	printf ("Inserting elements\n");
	do
	{
		insert2 = Malloc (sizeof (hntest2));
		printf ("   Insert: %s\n", *it);
		HASH_TABLE_NODE_SET_KEY(&(insert2->node),  *it);
		insert2->string2 = StringDup (*it);
		HashTableInsert (test1, insert2);
	}
	while (*(++it));

	it--;
	do
	{
		printf ("   Lookup: %s\n", *it);
		insert2 = HashTableLookup (test1, *it);
		printf ("   Verify: %s\n", insert2->string2);
		if (strcmp (insert2->string2, *it))
		{
			printf ("Failed!\n");
			exit (-1);
		}
	}
	while (*(--it));

	HashTableWalk (test1, SP2, NULL);

	printf ("Creating displaced hash table\n");
	/* We use small hash table size to allow thourough testing */
	test1 =
		HashTableNew (3, offsetof (hntest2, node), (t_hash_func) StringHash, (t_int32 (*) (void *, void *)) StringCmp, NULL);
	printf ("Done\n");

	it++;
	printf ("Prepend elements\n");
	do
	{
		insert2 = Malloc (sizeof (hntest2));
		printf ("   Prepend: %s\n", *it);
		HASH_TABLE_NODE_SET_KEY(&(insert2->node),  *it);
		insert2->string2 = StringDup (*it);
		HashTablePrepend (test1, insert2);
	}
	while (*(++it));

	it--;
	do
	{
		printf ("   Lookup: %s\n", *it);
		insert2 = HashTableLookup (test1, *it);
		printf ("   Verify: %s\n", insert2->string2);
		if (strcmp (insert2->string2, *it))
		{
			printf ("Failed!\n");
			exit (-1);
		}
	}
	while (*(--it));

	HashTableWalk (test1, SP2, NULL);


	return 0;
}
