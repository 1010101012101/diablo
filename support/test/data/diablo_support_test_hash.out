Creating simple hash table
Done
Inserting elements
   Insert: Search
   Insert: seek
   Insert: and
   Insert: know
   Insert: how
   Insert: this
   Insert: foul
   Insert: murder
   Insert: comes
   Insert: -----
   Insert: Who
   Insert: seeks
   Insert: and
   Insert: will
   Insert: not
   Insert: take
   Insert: when
   Insert: once
   Insert: 'tis
   Insert: offer'd
   Insert: Shall
   Insert: never
   Insert: find
   Insert: it
   Insert: more
   Insert: -----
   Insert: Search
   Insert: seek
   Insert: and
   Insert: know
   Insert: how
   Insert: this
   Insert: foul
   Insert: murder
   Insert: comes
   Lookup: comes
   Verify: comes
   Lookup: murder
   Verify: murder
   Lookup: foul
   Verify: foul
   Lookup: this
   Verify: this
   Lookup: how
   Verify: how
   Lookup: know
   Verify: know
   Lookup: and
   Verify: and
   Lookup: seek
   Verify: seek
   Lookup: Search
   Verify: Search
   Lookup: -----
   Verify: -----
   Lookup: more
   Verify: more
   Lookup: it
   Verify: it
   Lookup: find
   Verify: find
   Lookup: never
   Verify: never
   Lookup: Shall
   Verify: Shall
   Lookup: offer'd
   Verify: offer'd
   Lookup: 'tis
   Verify: 'tis
   Lookup: once
   Verify: once
   Lookup: when
   Verify: when
   Lookup: take
   Verify: take
   Lookup: not
   Verify: not
   Lookup: will
   Verify: will
   Lookup: and
   Verify: and
   Lookup: seeks
   Verify: seeks
   Lookup: Who
   Verify: Who
   Lookup: -----
   Verify: -----
   Lookup: comes
   Verify: comes
   Lookup: murder
   Verify: murder
   Lookup: foul
   Verify: foul
   Lookup: this
   Verify: this
   Lookup: how
   Verify: how
   Lookup: know
   Verify: know
   Lookup: and
   Verify: and
   Lookup: seek
   Verify: seek
   Lookup: Search
   Verify: Search
p->string=offer'd
p->string='tis
p->string=when
p->string=-----
p->string=-----
p->string=comes
p->string=comes
p->string=foul
p->string=foul
p->string=and
p->string=and
p->string=and
p->string=more
p->string=never
p->string=once
p->string=not
p->string=will
p->string=Who
p->string=murder
p->string=murder
p->string=this
p->string=this
p->string=how
p->string=how
p->string=seek
p->string=seek
p->string=it
p->string=find
p->string=Shall
p->string=take
p->string=seeks
p->string=know
p->string=know
p->string=Search
p->string=Search
Prepending elements
Creating simple hash table
Done
Inserting elements
   Prepend: Search
   Prepend: seek
   Prepend: and
   Prepend: know
   Prepend: how
   Prepend: this
   Prepend: foul
   Prepend: murder
   Prepend: comes
   Prepend: -----
   Prepend: Who
   Prepend: seeks
   Prepend: and
   Prepend: will
   Prepend: not
   Prepend: take
   Prepend: when
   Prepend: once
   Prepend: 'tis
   Prepend: offer'd
   Prepend: Shall
   Prepend: never
   Prepend: find
   Prepend: it
   Prepend: more
   Prepend: -----
   Prepend: Search
   Prepend: seek
   Prepend: and
   Prepend: know
   Prepend: how
   Prepend: this
   Prepend: foul
   Prepend: murder
   Prepend: comes
   Lookup: comes
   Verify: comes
   Lookup: murder
   Verify: murder
   Lookup: foul
   Verify: foul
   Lookup: this
   Verify: this
   Lookup: how
   Verify: how
   Lookup: know
   Verify: know
   Lookup: and
   Verify: and
   Lookup: seek
   Verify: seek
   Lookup: Search
   Verify: Search
   Lookup: -----
   Verify: -----
   Lookup: more
   Verify: more
   Lookup: it
   Verify: it
   Lookup: find
   Verify: find
   Lookup: never
   Verify: never
   Lookup: Shall
   Verify: Shall
   Lookup: offer'd
   Verify: offer'd
   Lookup: 'tis
   Verify: 'tis
   Lookup: once
   Verify: once
   Lookup: when
   Verify: when
   Lookup: take
   Verify: take
   Lookup: not
   Verify: not
   Lookup: will
   Verify: will
   Lookup: and
   Verify: and
   Lookup: seeks
   Verify: seeks
   Lookup: Who
   Verify: Who
   Lookup: -----
   Verify: -----
   Lookup: comes
   Verify: comes
   Lookup: murder
   Verify: murder
   Lookup: foul
   Verify: foul
   Lookup: this
   Verify: this
   Lookup: how
   Verify: how
   Lookup: know
   Verify: know
   Lookup: and
   Verify: and
   Lookup: seek
   Verify: seek
   Lookup: Search
   Verify: Search
p->string=offer'd
p->string='tis
p->string=when
p->string=-----
p->string=-----
p->string=comes
p->string=comes
p->string=foul
p->string=foul
p->string=and
p->string=and
p->string=and
p->string=more
p->string=never
p->string=once
p->string=not
p->string=will
p->string=Who
p->string=murder
p->string=murder
p->string=this
p->string=this
p->string=how
p->string=how
p->string=seek
p->string=seek
p->string=it
p->string=find
p->string=Shall
p->string=take
p->string=seeks
p->string=know
p->string=know
p->string=Search
p->string=Search
Creating displaced hash table
Done
Inserting elements
   Insert: Search
   Insert: seek
   Insert: and
   Insert: know
   Insert: how
   Insert: this
   Insert: foul
   Insert: murder
   Insert: comes
   Insert: -----
   Insert: Who
   Insert: seeks
   Insert: and
   Insert: will
   Insert: not
   Insert: take
   Insert: when
   Insert: once
   Insert: 'tis
   Insert: offer'd
   Insert: Shall
   Insert: never
   Insert: find
   Insert: it
   Insert: more
   Insert: -----
   Insert: Search
   Insert: seek
   Insert: and
   Insert: know
   Insert: how
   Insert: this
   Insert: foul
   Insert: murder
   Insert: comes
   Lookup: comes
   Verify: comes
   Lookup: murder
   Verify: murder
   Lookup: foul
   Verify: foul
   Lookup: this
   Verify: this
   Lookup: how
   Verify: how
   Lookup: know
   Verify: know
   Lookup: and
   Verify: and
   Lookup: seek
   Verify: seek
   Lookup: Search
   Verify: Search
   Lookup: -----
   Verify: -----
   Lookup: more
   Verify: more
   Lookup: it
   Verify: it
   Lookup: find
   Verify: find
   Lookup: never
   Verify: never
   Lookup: Shall
   Verify: Shall
   Lookup: offer'd
   Verify: offer'd
   Lookup: 'tis
   Verify: 'tis
   Lookup: once
   Verify: once
   Lookup: when
   Verify: when
   Lookup: take
   Verify: take
   Lookup: not
   Verify: not
   Lookup: will
   Verify: will
   Lookup: and
   Verify: and
   Lookup: seeks
   Verify: seeks
   Lookup: Who
   Verify: Who
   Lookup: -----
   Verify: -----
   Lookup: comes
   Verify: comes
   Lookup: murder
   Verify: murder
   Lookup: foul
   Verify: foul
   Lookup: this
   Verify: this
   Lookup: how
   Verify: how
   Lookup: know
   Verify: know
   Lookup: and
   Verify: and
   Lookup: seek
   Verify: seek
   Lookup: Search
   Verify: Search
p->string=offer'd
p->string='tis
p->string=when
p->string=-----
p->string=-----
p->string=comes
p->string=comes
p->string=foul
p->string=foul
p->string=and
p->string=and
p->string=and
p->string=more
p->string=never
p->string=once
p->string=not
p->string=will
p->string=Who
p->string=murder
p->string=murder
p->string=this
p->string=this
p->string=how
p->string=how
p->string=seek
p->string=seek
p->string=it
p->string=find
p->string=Shall
p->string=take
p->string=seeks
p->string=know
p->string=know
p->string=Search
p->string=Search
Creating displaced hash table
Done
Prepend elements
   Prepend: Search
   Prepend: seek
   Prepend: and
   Prepend: know
   Prepend: how
   Prepend: this
   Prepend: foul
   Prepend: murder
   Prepend: comes
   Prepend: -----
   Prepend: Who
   Prepend: seeks
   Prepend: and
   Prepend: will
   Prepend: not
   Prepend: take
   Prepend: when
   Prepend: once
   Prepend: 'tis
   Prepend: offer'd
   Prepend: Shall
   Prepend: never
   Prepend: find
   Prepend: it
   Prepend: more
   Prepend: -----
   Prepend: Search
   Prepend: seek
   Prepend: and
   Prepend: know
   Prepend: how
   Prepend: this
   Prepend: foul
   Prepend: murder
   Prepend: comes
   Lookup: comes
   Verify: comes
   Lookup: murder
   Verify: murder
   Lookup: foul
   Verify: foul
   Lookup: this
   Verify: this
   Lookup: how
   Verify: how
   Lookup: know
   Verify: know
   Lookup: and
   Verify: and
   Lookup: seek
   Verify: seek
   Lookup: Search
   Verify: Search
   Lookup: -----
   Verify: -----
   Lookup: more
   Verify: more
   Lookup: it
   Verify: it
   Lookup: find
   Verify: find
   Lookup: never
   Verify: never
   Lookup: Shall
   Verify: Shall
   Lookup: offer'd
   Verify: offer'd
   Lookup: 'tis
   Verify: 'tis
   Lookup: once
   Verify: once
   Lookup: when
   Verify: when
   Lookup: take
   Verify: take
   Lookup: not
   Verify: not
   Lookup: will
   Verify: will
   Lookup: and
   Verify: and
   Lookup: seeks
   Verify: seeks
   Lookup: Who
   Verify: Who
   Lookup: -----
   Verify: -----
   Lookup: comes
   Verify: comes
   Lookup: murder
   Verify: murder
   Lookup: foul
   Verify: foul
   Lookup: this
   Verify: this
   Lookup: how
   Verify: how
   Lookup: know
   Verify: know
   Lookup: and
   Verify: and
   Lookup: seek
   Verify: seek
   Lookup: Search
   Verify: Search
p->string=offer'd
p->string='tis
p->string=when
p->string=-----
p->string=-----
p->string=comes
p->string=comes
p->string=foul
p->string=foul
p->string=and
p->string=and
p->string=and
p->string=more
p->string=never
p->string=once
p->string=not
p->string=will
p->string=Who
p->string=murder
p->string=murder
p->string=this
p->string=this
p->string=how
p->string=how
p->string=seek
p->string=seek
p->string=it
p->string=find
p->string=Shall
p->string=take
p->string=seeks
p->string=know
p->string=know
p->string=Search
p->string=Search
